saq/queue/postgres.py

```diff
@@ -0,0 +1,648 @@
+                    channel=Identifier(channel), payload=payload
+                )
+            )
+
+    async def update(
+        self,
+        job: Job,
+        status: Status | None = None,
+        scheduled: int | None = None,
+        ttl: float | None = None,
+        connection: AsyncConnection | None = None,
+    ) -> None:
+        if status:
+            job.status = status
+        if scheduled:
+            job.scheduled = scheduled
+        job.touched = now()
+
+        async with (
+            self.nullcontext(connection) if connection else self.pool.connection()
+        ) as conn:
+            await conn.execute(
+                SQL(
+                    """
+                    UPDATE {jobs_table} SET job = %(job)s, status = %(status)s, scheduled = %(scheduled)s, ttl = %(ttl)s
+                    WHERE key = %(key)s
+                    """
+                ).format(jobs_table=self.jobs_table),
+                {
+                    "job": self.serialize(job),
+                    "status": job.status,
+                    "key": job.key,
+                    "scheduled": job.scheduled,
+                    "ttl": ttl,
+                },
+            )
+            await self.notify(job, conn)
+
+    async def job(self, job_key: str) -> Job | None:
+        async with self.pool.connection() as conn, conn.cursor() as cursor:
+            await cursor.execute(
+                SQL(
+                    """
+                SELECT job
+                FROM {jobs_table}
+                WHERE key = %(key)s
+                """
+                ).format(jobs_table=self.jobs_table),
+                {"key": job_key},
+            )
+            job = await cursor.fetchone()
+            if job:
+                return self.deserialize(job[0])
+        return None
+
+    async def abort(self, job: Job, error: str, ttl: float = 5) -> None:
+        job.error = error
+        job.status = Status.ABORTING
+        job.ttl = int(seconds(now()) + ttl) + 1
+
+        await self.update(job)
+
+    async def retry(self, job: Job, error: str | None) -> None:
+        self._update_job_for_retry(job, error)
+        next_retry_delay = job.next_retry_delay()
+        if next_retry_delay:
+            scheduled = time.time() + next_retry_delay
+        else:
+            scheduled = job.scheduled or seconds(now())
+
+        await self.update(job, scheduled=int(scheduled))
+        await self._release_job(job.key)
+
+        self.retried += 1
+        await self.notify(job)
+        logger.info("Retrying %s", job.info(logger.isEnabledFor(logging.DEBUG)))
+
+    async def finish(
+        self,
+        job: Job,
+        status: Status,
+        *,
+        result: t.Any = None,
+        error: str | None = None,
+    ) -> None:
+        self._update_job_for_finish(job, status, result=result, error=error)
+        key = job.key
+
+        async with self.pool.connection() as conn, conn.cursor() as cursor:
+            if job.ttl >= 0:
```

Comment:
hey, this doesn't match redis's behavior, if ttl > 0, then set expiration, if it's 0, there is no expiration, else delete. right?

Reply:
Good catch. This is wrong after I refactored finish() to use update(). I've fixed it.